# Cleaning and Analysing a Dataset of Used Cars from eBay

In this project, we'll work with a modified dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The data dictionary provided with data is as follows:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest`- Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `odometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('C:\\Users\\Admin\\Desktop\\Dataset\\autos.csv')

We will now do some exploring thourgh the dataset.

In [2]:
autos.info()
autos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

**As we can see, only five columns have null data:**
* vehicleType
* gearbox
* model
* fuelType
* notRepairedDamage

**And five columns stored data as integers:**
* yearOfRegistration
* powerPS 
* monthOfRegistration
* nrOfPictures 
* postalCode 

Because the column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

 We need to change the name of the columns by using the `Dataframe.columns` attribute.

In [4]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test", "vehicle_type", 
                "registration_year", "gearbox", "power_ps", "model", "odometer", "registration_month", "fuel_type", "brand",
                "unrepaired_damage", "ad_created", "num_photos", "postal_code", "last_seen"]

In [5]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

**As you can see, the columns were renamed.**

Because this dataset is from a German website, maybe some columns will contain German words. We will explore and deal with it later.

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
* Examples of numeric data stored as text which can be cleaned and converted.


In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

The `price` and `odometer`columns are numeric values stored as text. So, for each columns: 
* We will remove any non-numeric characters.
* Convert the columns to a numeric dtype.
* Finally, rename the `odometer` column to `odometer_km` so it is clear what the columns's values are about.

Based on the frequency of the values, we can know that the `seller`, `offer_type` and `num_photos` columns have mostly one value each. As it does not make sense to have columns like this in the dataset, we will start the cleaning process by dropping those columns.

Moving on, we can see that `price` and `odometer_km` data is stored as string because of characters such as '**,**', '**$**' and '**km**'. Follow the steps mentioned above, we will clean and convert these. 

In [7]:
autos["price"] = (autos["price"].str.replace("$", "")
                  .str.replace(",", "")
                  .astype(int))
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [8]:
autos["odometer"] = (autos["odometer"].str.replace("km", "")
                     .str.replace(",", "")
                     .astype(int))

autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

Now let's take a closer look in these two columns, starting with `odometer_km`.

In [9]:
# check that how many unique values
print(autos["odometer_km"].unique().shape, "\n")

# view min/max/median/mean etc
print(autos["odometer_km"].describe(), "\n")

# unique value in odometer column
uni_value_odo = autos["odometer_km"].sort_values(ascending = False).unique()
print(uni_value_odo)

(13,) 

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

[150000 125000 100000  90000  80000  70000  60000  50000  40000  30000
  20000  10000   5000]


It looks like the values have been rounded. The highest value is 150,000 and the lowest value is 5000. This seems plausible.

Next, let's check the `price` column.

In [20]:
# check that how many unique values
print(autos["price"].unique().shape, "\n")

# view min/max/median/mean etc
print(autos["price"].describe(), "\n")

# unique value in odometer column
uni_value_price = autos["price"].sort_values(ascending = False).unique()
print(uni_value_price)

(2357,) 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

[99999999 27322222 12345678 ...        2        1        0]


The smallest value is 0 and the highest value is 99,999,999, while the second highest value is 27,322,222. The difference in value is too great, it seems that the data is not very reliable. Therefore, we need to eliminate data that is deemed unrealistic. 

We need data in a realistic range of values that can be analyzed on it. Let's dig deeper to see if we can find anything.

In [21]:
print(autos['price'].value_counts().sort_index(ascending=False).head(20), "\n")

print(autos['price'].value_counts().sort_index(ascending=True).head(20))

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64 

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64


We now can see the data more clearly. To make the analysis as close to reality as possible, we'll exclude rows where the `price` column is above 100 and below $1,000,000.

In [22]:
autos = autos[autos['price'].between(100, 1000000)]
autos['price'].describe()

count     48227.000000
mean       5992.208099
std       11994.952272
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      999999.000000
Name: price, dtype: float64

Now the data has become more realistic.

**The next thing to do is remove unnecessary columns.**

Before removing those columns, let's take a look into them.

In [10]:
print(autos["seller"].value_counts(), "\n")
print(autos["offer_type"].value_counts(), "\n")
print(autos["num_photos"].value_counts(), "\n")

seller
privat        49999
gewerblich        1
Name: count, dtype: int64 

offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64 

num_photos
0    50000
Name: count, dtype: int64 



The `seller` and `offer_type` columns have a row with a value different from the others. 

And the `num_photos` columns only have one value.

We will now drop those columns and check if the columns were actually deleted.

In [11]:
print(autos.shape)
autos = autos.drop(["seller", "offer_type", "num_photos"], axis = 1)
print(autos.shape)

(50000, 20)
(50000, 17)


Next, we will deal with the previous problem: columns containing German words. Based on what we discorvered from the dataset, we found that `gearbox`, `fuel_type` and `unrepaired_damage` are columns cotaining German words.

In [14]:
print(autos["gearbox"].unique(), "\n")
print(autos["fuel_type"].unique(), "\n")
print(autos["unrepaired_damage"].unique(), "\n")

['manuell' 'automatik' nan] 

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere'] 

['nein' nan 'ja'] 



We need to translate these words to English, and then replace it to our dataset.

We will do it in the following steps:
* We create a dictionary for each column, where German words are the keys, and English words are the values.
* After that, we'll use the `Series.map` to replace the German words for the English words in each column for the whole dataset. 

In [15]:
dict_gearbox = {'manuell' : 'manually', 'automatik' : 'automatic'}
dict_fuel_type = {'lpg' : 'lpg', 'benzin' : 'gasoline', 'diesel' : 'diesel', 'cng' : 'cng', 
                'hybrid' : 'hybrid', 'elektro' : 'elektro', 'andere' : 'other'}
dict_unrepaired_dmg = {'nein' : 'no', 'ja': 'yes'}

autos["gearbox"] = autos["gearbox"].map(dict_gearbox)
autos["fuel_type"] = autos["fuel_type"].map(dict_fuel_type)
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(dict_unrepaired_dmg)

Check if the columns' data has been replaced. 

In [17]:
print(autos.loc[1:5, ["gearbox", "fuel_type", "unrepaired_damage"]])

     gearbox fuel_type unrepaired_damage
1  automatic  gasoline                no
2   manually  gasoline                no
3  automatic  gasoline                no
4   manually  gasoline                no
5  automatic    diesel               NaN


The problem has been resolved. 